File: rust/src/cmd/list.rs
use std::path::Path;

use clap::Args;
use serde::Serialize;

use crate::output::OutputFormat;
use crate::thread::{self, Thread};
use crate::workspace;

#[derive(Args)]
pub struct ListArgs {
    /// Path to list threads from (git-root-relative, ./pwd-relative, or absolute)
    #[arg(default_value = "")]
    path: String,

    /// Include nested directories (recursive search)
    #[arg(short = 'r', long)]
    recursive: bool,

    /// Include resolved/terminal threads
    #[arg(long)]
    include_closed: bool,

    /// Search name/title/desc (substring)
    #[arg(short = 's', long)]
    search: Option<String>,

    /// Filter by status (comma-separated)
    #[arg(long)]
    status: Option<String>,

    /// Output format (auto-detects TTY for fancy vs plain)
    #[arg(short = 'f', long, value_enum, default_value = "fancy")]
    format: OutputFormat,

    /// Output as JSON (shorthand for --format=json)
    #[arg(long, conflicts_with = "format")]
    json: bool,
}

#[derive(Serialize)]
struct ThreadInfo {
    id: String,
    status: String,
    path: String,
    name: String,
    title: String,
    desc: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    path_absolute: Option<String>,
    #[serde(skip_serializing_if = "std::ops::Not::not")]
    is_pwd: bool,
}

pub fn run(args: ListArgs, git_root: &Path) -> Result<(), String> {
    // Determine output format (handle --json shorthand)
    let format = if args.json {
        OutputFormat::Json
    } else {
        args.format.resolve()
    };

    // Parse path filter if provided
    let path_filter = if args.path.is_empty() {
        None
    } else {
        Some(args.path.as_str())
    };

    // Resolve the scope to understand current context
    let scope = workspace::infer_scope(git_root, path_filter)?;
    let filter_path = scope.path.clone();

    let threads = workspace::find_all_threads(git_root)?;
    let mut results = Vec::new();

    // Get PWD relative path for comparison
    let pwd_rel = workspace::pwd_relative_to_git_root(git_root).unwrap_or_else(|_| ".".to_string());

    // Determine if we need absolute paths (for json/yaml)
    let include_absolute = matches!(format, OutputFormat::Json | OutputFormat::Yaml);

    for thread_path in threads {
        let t = match Thread::parse(&thread_path) {
            Ok(t) => t,
            Err(_) => continue,
        };

        let rel_path = workspace::parse_thread_path(git_root, &thread_path);
        let status = t.status().to_string();
        let base_status = thread::base_status(&status);
        let name = thread::extract_name_from_path(&thread_path);

        // Path filter: if not recursive, only show threads at the specified level
        if !args.recursive {
            // At specific path: only show threads in that exact directory
            if rel_path != filter_path {
                continue;
            }
        } else {
            // Recursive mode: show threads at or under the filter path
            if filter_path != "." {
                let filter_prefix = if filter_path.ends_with('/') {
                    filter_path.clone()
                } else {
                    format!("{}/", filter_path)
                };
                if rel_path != filter_path && !rel_path.starts_with(&filter_prefix) {
                    continue;
                }
            }
        }

        // Status filter
        if let Some(ref status_filter) = args.status {
            let filter_statuses: Vec<&str> = status_filter.split(',').collect();
            if !filter_statuses.contains(&base_status.as_str()) {
                continue;
            }
        } else if !args.include_closed && thread::is_terminal(&status) {
            continue;
        }

        // Search filter
        if let Some(ref search) = args.search {
            let search_lower = search.to_lowercase();
            let name_lower = name.to_lowercase();
            let title_lower = t.name().to_lowercase();
            let desc_lower = t.frontmatter.desc.to_lowercase();

            if !name_lower.contains(&search_lower)
                && !title_lower.contains(&search_lower)
                && !desc_lower.contains(&search_lower)
            {
                continue;
            }
        }

        // Use title if available, else humanize name
        let title = if !t.name().is_empty() {
            t.name().to_string()
        } else {
            name.replace('-', " ")
        };

        let is_pwd = rel_path == pwd_rel;

        results.push(ThreadInfo {
            id: t.id().to_string(),
            status: base_status,
            path: rel_path,
            name,
            title,
            desc: t.frontmatter.desc.clone(),
            path_absolute: if include_absolute {
                Some(thread_path.to_string_lossy().to_string())
            } else {
                None
            },
            is_pwd,
        });
    }

    match format {
        OutputFormat::Fancy => output_fancy(
            &results,
            git_root,
            &filter_path,
            &pwd_rel,
            args.recursive,
            args.include_closed,
            args.status.as_deref(),
        ),
        OutputFormat::Plain => output_plain(
            &results,
            git_root,
            &filter_path,
            &pwd_rel,
            args.recursive,
            args.include_closed,
            args.status.as_deref(),
        ),
        OutputFormat::Json => output_json(&results, git_root, &pwd_rel),
        OutputFormat::Yaml => output_yaml(&results, git_root, &pwd_rel),
    }
}

fn output_fancy(
    results: &[ThreadInfo],
    git_root: &Path,
    filter_path: &str,
    pwd_rel: &str,
    recursive: bool,
    include_closed: bool,
    status_filter: Option<&str>,
) -> Result<(), String> {
    // Fancy header: repo-name (rel/path/to/pwd)
    let repo_name = git_root
        .file_name()
        .map(|n| n.to_string_lossy().to_string())
        .unwrap_or_else(|| "repo".to_string());

    let path_desc = if filter_path == "." {
        String::new()
    } else {
        format!(" ({})", filter_path)
    };

    let pwd_marker = if filter_path == pwd_rel { " ← PWD" } else { "" };

    println!("{}{}{}", repo_name, path_desc, pwd_marker);
    println!();

    let status_desc = if let Some(s) = status_filter {
        format!("{} ", s)
    } else if include_closed {
        String::new()
    } else {
        "active ".to_string()
    };

    let recursive_suffix = if recursive { " (recursive)" } else { "" };

    println!(
        "Showing {} {}threads{}",
        results.len(),
        status_desc,
        recursive_suffix
    );
    println!();

    if results.is_empty() {
        if !recursive {
            println!("Hint: use -r to include nested directories");
        }
        return Ok(());
    }

    // Print table header
    println!(
        "{:<6} {:<10} {:<24} {}",
        "ID", "STATUS", "PATH", "NAME"
    );
    println!(
        "{:<6} {:<10} {:<24} {}",
        "--", "------", "----", "----"
    );

    for t in results {
        let path_display = truncate(&t.path, 22);
        let pwd_marker = if t.is_pwd { " ←" } else { "" };
        println!(
            "{:<6} {:<10} {:<24} {}{}",
            t.id, t.status, path_display, t.title, pwd_marker
        );
    }

    Ok(())
}

fn output_plain(
    results: &[ThreadInfo],
    git_root: &Path,
    filter_path: &str,
    pwd_rel: &str,
    recursive: bool,
    include_closed: bool,
    status_filter: Option<&str>,
) -> Result<(), String> {
    // Plain header: explicit context
    let pwd = std::env::current_dir()
        .map(|p| p.to_string_lossy().to_string())
        .unwrap_or_else(|_| "(unknown)".to_string());

    println!("PWD: {}", pwd);
    println!("Git root: {}", git_root.display());
    println!("PWD (git-relative): {}", pwd_rel);
    println!();

    let path_desc = if filter_path == "." {
        "repo root".to_string()
    } else {
        filter_path.to_string()
    };

    let status_desc = if let Some(s) = status_filter {
        s.to_string()
    } else if include_closed {
        String::new()
    } else {
        "active".to_string()
    };

    let recursive_suffix = if recursive { " (recursive)" } else { "" };

    let pwd_suffix = if filter_path == pwd_rel {
        " ← PWD"
    } else {
        ""
    };

    if !status_desc.is_empty() {
        println!(
            "Showing {} {} threads in {}{}{}",
            results.len(),
            status_desc,
            path_desc,
            recursive_suffix,
            pwd_suffix
        );
    } else {
        println!(
            "Showing {} threads in {} (all statuses){}{}",
            results.len(),
            path_desc,
            recursive_suffix,
            pwd_suffix
        );
    }
    println!();

    if results.is_empty() {
        if !recursive {
            println!("Hint: use -r to include nested directories");
        }
        return Ok(());
    }

    // Print table header
    println!(
        "{:<6} {:<10} {:<24} {}",
        "ID", "STATUS", "PATH", "NAME"
    );
    println!(
        "{:<6} {:<10} {:<24} {}",
        "--", "------", "----", "----"
    );

    for t in results {
        let path_display = truncate(&t.path, 22);
        let pwd_marker = if t.is_pwd { " ← PWD" } else { "" };
        println!(
            "{:<6} {:<10} {:<24} {}{}",
            t.id, t.status, path_display, t.title, pwd_marker
        );
    }

    Ok(())
}

fn output_json(
    results: &[ThreadInfo],
    git_root: &Path,
    pwd_rel: &str,
) -> Result<(), String> {
    let pwd = std::env::current_dir()
        .map(|p| p.to_string_lossy().to_string())
        .unwrap_or_else(|_| String::new());

    #[derive(Serialize)]
    struct JsonOutput<'a> {
        pwd: String,
        git_root: String,
        pwd_relative: &'a str,
        threads: &'a [ThreadInfo],
    }

    let output = JsonOutput {
        pwd,
        git_root: git_root.to_string_lossy().to_string(),
        pwd_relative: pwd_rel,
        threads: results,
    };

    let json =
        serde_json::to_string_pretty(&output).map_err(|e| format!("JSON serialization failed: {}", e))?;
    println!("{}", json);
    Ok(())
}

fn output_yaml(
    results: &[ThreadInfo],
    git_root: &Path,
    pwd_rel: &str,
) -> Result<(), String> {
    let pwd = std::env::current_dir()
        .map(|p| p.to_string_lossy().to_string())
        .unwrap_or_else(|_| String::new());

    #[derive(Serialize)]
    struct YamlOutput<'a> {
        pwd: String,
        git_root: String,
        pwd_relative: &'a str,
        threads: &'a [ThreadInfo],
    }

    let output = YamlOutput {
        pwd,
        git_root: git_root.to_string_lossy().to_string(),
        pwd_relative: pwd_rel,
        threads: results,
    };

    let yaml =
        serde_yaml::to_string(&output).map_err(|e| format!("YAML serialization failed: {}", e))?;
    print!("{}", yaml);
    Ok(())
}

fn truncate(s: &str, max_chars: usize) -> String {
    let char_count = s.chars().count();
    if char_count <= max_chars {
        s.to_string()
    } else {
        let truncated: String = s.chars().take(max_chars - 1).collect();
        format!("{}…", truncated)
    }
}

File: go/internal/cmd/list.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/spf13/cobra"
	"gopkg.in/yaml.v3"

	"git.zib.de/cspiegel/threads/internal/output"
	"git.zib.de/cspiegel/threads/internal/thread"
	"git.zib.de/cspiegel/threads/internal/workspace"
)

var (
	listDown          *int
	listRecursive     bool
	listUp            *int
	listIncludeClosed bool
	listSearch        string
	listStatus        string
	listFormat        string
	listJSON          bool
)

var listCmd = &cobra.Command{
	Use:     "list [path]",
	Aliases: []string{"ls"},
	Short:   "List threads",
	Long: `List threads at the specified level.

Path resolution:
  (none)  → PWD (current directory)
  .       → PWD (explicit)
  ./X/Y   → PWD-relative
  /X/Y    → Absolute
  X/Y     → Git-root-relative

By default shows active threads at the current level only.
Use -d/--down to include subdirectories, -u/--up to include parent directories.
Use -r as an alias for --down (unlimited depth).
Use --include-closed to include resolved/terminal threads.`,
	Args: cobra.MaximumNArgs(1),
	RunE: runList,
}

var listDownVal int
var listUpVal int

func init() {
	listCmd.Flags().IntVarP(&listDownVal, "down", "d", -1, "Search subdirectories (N levels, 0=unlimited)")
	listCmd.Flags().BoolVarP(&listRecursive, "recursive", "r", false, "Alias for --down (unlimited depth)")
	listCmd.Flags().IntVarP(&listUpVal, "up", "u", -1, "Search parent directories (N levels, 0=to git root)")
	listCmd.Flags().BoolVar(&listIncludeClosed, "include-closed", false, "Include resolved/terminal threads")
	listCmd.Flags().StringVarP(&listSearch, "search", "s", "", "Search name/title/desc (substring)")
	listCmd.Flags().StringVar(&listStatus, "status", "", "Filter by status (comma-separated)")
	listCmd.Flags().StringVarP(&listFormat, "format", "f", "fancy", "Output format: fancy, plain, json, yaml")
	listCmd.Flags().BoolVar(&listJSON, "json", false, "Output as JSON (shorthand for --format=json)")
}

// searchDirection describes the search direction for output display.
type searchDirection struct {
	hasDown   bool
	downDepth int // -1 = unlimited, 0+ = specific depth
	hasUp     bool
	upDepth   int // -1 = unlimited (to git root), 0+ = specific depth
}

func (s *searchDirection) description() string {
	var parts []string

	if s.hasDown {
		if s.downDepth < 0 {
			parts = append(parts, "recursive")
		} else {
			parts = append(parts, fmt.Sprintf("down %d", s.downDepth))
		}
	}

	if s.hasUp {
		if s.upDepth < 0 {
			parts = append(parts, "up")
		} else {
			parts = append(parts, fmt.Sprintf("up %d", s.upDepth))
		}
	}

	if len(parts) == 0 {
		return ""
	}
	return fmt.Sprintf(" (%s)", strings.Join(parts, ", "))
}

func (s *searchDirection) isSearching() bool {
	return s.hasDown || s.hasUp
}

type threadInfo struct {
	ID           string `json:"id" yaml:"id"`
	Status       string `json:"status" yaml:"status"`
	Path         string `json:"path" yaml:"path"`
	Name         string `json:"name" yaml:"name"`
	Title        string `json:"title" yaml:"title"`
	Desc         string `json:"desc" yaml:"desc"`
	PathAbsolute string `json:"path_absolute,omitempty" yaml:"path_absolute,omitempty"`
	IsPwd        bool   `json:"is_pwd,omitempty" yaml:"is_pwd,omitempty"`
}

func runList(cmd *cobra.Command, args []string) error {
	gitRoot := getWorkspace()

	// Determine output format (handle --json shorthand)
	var format output.Format
	if listJSON {
		format = output.FormatJSON
	} else {
		format, _ = output.ParseFormat(listFormat)
		format = format.Resolve()
	}

	// Parse path filter if provided
	pathArg := ""
	if len(args) > 0 {
		pathArg = args[0]
	}

	// Resolve the scope
	scope, err := workspace.InferScope(gitRoot, pathArg)
	if err != nil {
		return err
	}
	filterPath := scope.Path
	startPath := filepath.Dir(scope.ThreadsDir)

	// Determine search direction: --down/-d takes priority, then -r as alias
	downSet := cmd.Flags().Changed("down")
	hasDown := downSet || listRecursive
	downDepth := -1 // unlimited by default
	if downSet && listDownVal > 0 {
		downDepth = listDownVal
	}

	upSet := cmd.Flags().Changed("up")
	hasUp := upSet
	upDepth := -1 // unlimited (to git root) by default
	if upSet && listUpVal > 0 {
		upDepth = listUpVal
	}

	// Build options
	options := workspace.NewFindOptions()

	if hasDown {
		depth := downDepth
		if depth < 0 {
			depth = 0 // 0 means unlimited in our convention
		}
		options = options.WithDown(&depth)
	}

	if hasUp {
		depth := upDepth
		if depth < 0 {
			depth = 0 // 0 means unlimited in our convention
		}
		options = options.WithUp(&depth)
	}

	// Track search direction for output
	searchDir := &searchDirection{
		hasDown:   hasDown,
		downDepth: downDepth,
		hasUp:     hasUp,
		upDepth:   upDepth,
	}

	// Find threads using options
	threads, err := workspace.FindThreadsWithOptions(startPath, gitRoot, options)
	if err != nil {
		return err
	}

	// Get PWD relative path for comparison
	pwdRel, _ := workspace.PWDRelativeToGitRoot(gitRoot)

	// Determine if we need absolute paths (for json/yaml)
	includeAbsolute := format == output.FormatJSON || format == output.FormatYAML

	var results []threadInfo

	for _, path := range threads {
		t, err := thread.Parse(path)
		if err != nil {
			fmt.Fprintf(os.Stderr, "warning: failed to parse %s: %v\n", path, err)
			continue
		}

		relPath := workspace.ParseThreadPath(gitRoot, path)
		status := t.Status()
		baseStatus := thread.BaseStatus(status)
		name := thread.ExtractNameFromPath(path)

		// Path filter: if not searching, only show threads at the specified level
		if !searchDir.isSearching() {
			if relPath != filterPath {
				continue
			}
		}
		// Note: FindThreadsWithOptions already handles direction/depth filtering

		// Status filter
		statusFlagSet := cmd.Flags().Changed("status")
		if statusFlagSet {
			if listStatus == "" {
				continue
			}
			if !strings.Contains(","+listStatus+",", ","+baseStatus+",") {
				continue
			}
		} else {
			if !listIncludeClosed && thread.IsTerminal(status) {
				continue
			}
		}

		// Search filter
		if listSearch != "" {
			searchLower := strings.ToLower(listSearch)
			nameLower := strings.ToLower(name)
			titleLower := strings.ToLower(t.Name())
			descLower := strings.ToLower(t.Frontmatter.Desc)

			if !strings.Contains(nameLower, searchLower) &&
				!strings.Contains(titleLower, searchLower) &&
				!strings.Contains(descLower, searchLower) {
				continue
			}
		}

		// Use title if available, else humanize name
		title := t.Name()
		if title == "" {
			title = strings.ReplaceAll(name, "-", " ")
		}

		isPwd := relPath == pwdRel

		info := threadInfo{
			ID:     t.ID(),
			Status: baseStatus,
			Path:   relPath,
			Name:   name,
			Title:  title,
			Desc:   t.Frontmatter.Desc,
			IsPwd:  isPwd,
		}

		if includeAbsolute {
			info.PathAbsolute = path
		}

		results = append(results, info)
	}

	switch format {
	case output.FormatFancy:
		return outputFancy(results, gitRoot, filterPath, pwdRel, searchDir)
	case output.FormatPlain:
		return outputPlain(results, gitRoot, filterPath, pwdRel, searchDir)
	case output.FormatJSON:
		return outputJSON(results, gitRoot, pwdRel)
	case output.FormatYAML:
		return outputYAML(results, gitRoot, pwdRel)
	default:
		return outputFancy(results, gitRoot, filterPath, pwdRel, searchDir)
	}
}

func outputFancy(results []threadInfo, gitRoot, filterPath, pwdRel string, searchDir *searchDirection) error {
	// Fancy header: repo-name (rel/path/to/pwd)
	repoName := filepath.Base(gitRoot)

	pathDesc := ""
	if filterPath != "." {
		pathDesc = fmt.Sprintf(" (%s)", filterPath)
	}

	pwdMarker := ""
	if filterPath == pwdRel {
		pwdMarker = " ← PWD"
	}

	fmt.Printf("%s%s%s\n", repoName, pathDesc, pwdMarker)
	fmt.Println()

	statusDesc := "active "
	if listStatus != "" {
		statusDesc = listStatus + " "
	} else if listIncludeClosed {
		statusDesc = ""
	}

	searchSuffix := searchDir.description()

	fmt.Printf("Showing %d %sthreads%s\n", len(results), statusDesc, searchSuffix)
	fmt.Println()

	if len(results) == 0 {
		if !searchDir.isSearching() {
			fmt.Println("Hint: use -r to include nested directories, -u to search parents")
		}
		return nil
	}

	// Print table header
	fmt.Printf("%-6s %-10s %-24s %s\n", "ID", "STATUS", "PATH", "NAME")
	fmt.Printf("%-6s %-10s %-24s %s\n", "--", "------", "----", "----")

	for _, t := range results {
		pathDisplay := truncate(t.Path, 22)
		marker := ""
		if t.IsPwd {
			marker = " ←"
		}
		fmt.Printf("%-6s %-10s %-24s %s%s\n", t.ID, t.Status, pathDisplay, t.Title, marker)
	}

	return nil
}

func outputPlain(results []threadInfo, gitRoot, filterPath, pwdRel string, searchDir *searchDirection) error {
	// Plain header: explicit context
	pwd, _ := os.Getwd()
	fmt.Printf("PWD: %s\n", pwd)
	fmt.Printf("Git root: %s\n", gitRoot)
	fmt.Printf("PWD (git-relative): %s\n", pwdRel)
	fmt.Println()

	pathDesc := filterPath
	if filterPath == "." {
		pathDesc = "repo root"
	}

	statusDesc := "active"
	if listStatus != "" {
		statusDesc = listStatus
	} else if listIncludeClosed {
		statusDesc = ""
	}

	searchSuffix := searchDir.description()

	pwdSuffix := ""
	if filterPath == pwdRel {
		pwdSuffix = " ← PWD"
	}

	if statusDesc != "" {
		fmt.Printf("Showing %d %s threads in %s%s%s\n", len(results), statusDesc, pathDesc, searchSuffix, pwdSuffix)
	} else {
		fmt.Printf("Showing %d threads in %s (all statuses)%s%s\n", len(results), pathDesc, searchSuffix, pwdSuffix)
	}
	fmt.Println()

	if len(results) == 0 {
		if !searchDir.isSearching() {
			fmt.Println("Hint: use -r to include nested directories, -u to search parents")
		}
		return nil
	}

	// Print table header
	fmt.Printf("%-6s %-10s %-24s %s\n", "ID", "STATUS", "PATH", "NAME")
	fmt.Printf("%-6s %-10s %-24s %s\n", "--", "------", "----", "----")

	for _, t := range results {
		pathDisplay := truncate(t.Path, 22)
		pwdMarker := ""
		if t.IsPwd {
			pwdMarker = " ← PWD"
		}
		fmt.Printf("%-6s %-10s %-24s %s%s\n", t.ID, t.Status, pathDisplay, t.Title, pwdMarker)
	}

	return nil
}

func outputJSON(results []threadInfo, gitRoot, pwdRel string) error {
	pwd, _ := os.Getwd()

	type jsonOutput struct {
		PWD         string       `json:"pwd"`
		GitRoot     string       `json:"git_root"`
		PwdRelative string       `json:"pwd_relative"`
		Threads     []threadInfo `json:"threads"`
	}

	output := jsonOutput{
		PWD:         pwd,
		GitRoot:     gitRoot,
		PwdRelative: pwdRel,
		Threads:     results,
	}

	enc := json.NewEncoder(os.Stdout)
	enc.SetIndent("", "  ")
	return enc.Encode(output)
}

func outputYAML(results []threadInfo, gitRoot, pwdRel string) error {
	pwd, _ := os.Getwd()

	type yamlOutput struct {
		PWD         string       `yaml:"pwd"`
		GitRoot     string       `yaml:"git_root"`
		PwdRelative string       `yaml:"pwd_relative"`
		Threads     []threadInfo `yaml:"threads"`
	}

	output := yamlOutput{
		PWD:         pwd,
		GitRoot:     gitRoot,
		PwdRelative: pwdRel,
		Threads:     results,
	}

	enc := yaml.NewEncoder(os.Stdout)
	enc.SetIndent(2)
	return enc.Encode(output)
}

func truncate(s string, max int) string {
	if len(s) <= max {
		return s
	}
	return s[:max-1] + "…"
}
